corehq/apps/dhis2/api.py

```diff
@@ -0,0 +1,132 @@
+import json
+import logging
+import requests
+from corehq.apps.dhis2.dbaccessors import get_dhis2_connection
+
+
+logger = logging.getLogger('json_api_logger')
+
+
+class JsonApiError(Exception):
+    """
+    JsonApiError is raised for HTTP or socket errors.
+    """
+    pass
+
+
+def json_serializer(obj):
+    """
+    A JSON serializer that serializes dates and times
+    """
+    if hasattr(obj, 'isoformat'):
+        return obj.isoformat()
+
+
+def log_request(func):
+
+    def log(log_level, json_api_request, request_error, response_status, response_body, method_func, path,
+            data=None, **params):
+        """
+        Unpack function, path and data from args and kwargs in order to log them separately
+        """
+        logger.log(log_level, {
+            'domain': json_api_request.domain_name,
+            'log_level': log_level,
+            'request_method': method_func.__name__.upper(),
+            'request_url': json_api_request.server_url + path,
+            'request_headers': json.dumps(json_api_request.headers),
+            'request_params': json.dumps(params),
+            'request_body': '' if data is None else json.dumps(data, default=json_serializer),
+
+            'request_error': request_error,
+            'response_status': response_status,
+            'response_body': response_body,
+        })
```

Comment:
Can you articulate why you think this indirection is necessary / good? You could as easily call a function to save the model.